// type-mapping/dist/functional-lib/null/cast.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var type_util_1 = require("../../type-util");
var operator_1 = require("../operator");
var literal_1 = require("../literal");
var null_1 = require("./null");
var string_1 = require("../string");
var error_util_1 = require("../../error-util");
function undefinedToNull() {
    return function (name, mixed) {
        //We do not mind === here.
        if (mixed === undefined || mixed === null) {
            return null;
        }
        throw error_util_1.makeMappingError({
            message: name + " must be null|undefined; received " + type_util_1.toTypeStr(mixed),
            inputName: name,
            actualValue: mixed,
            expected: "null|undefined",
            expectedMeta: {
                mappableValues: [null, undefined],
                outputValues: [null],
            },
        });
    };
}
exports.undefinedToNull = undefinedToNull;
function emptyStringToNull() {
    return operator_1.cast(literal_1.literal(""), function () { return null; }, null_1.null());
}
exports.emptyStringToNull = emptyStringToNull;
/**
    An empty string, or a string of only whitespace
*/
function whitespaceStringToNull() {
    return operator_1.cast(string_1.match(/^\s*$/, function (name) {
        return {
            message: name + " must be a whitespace string",
            expected: "whitespace string",
        };
    }), function () { return null; }, null_1.null());
}
exports.whitespaceStringToNull = whitespaceStringToNull;
