// type-mapping/dist/functional-lib/date/instance-of-date.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var type_util_1 = require("../../type-util");
var number_1 = require("../number");
var error_util_1 = require("../../error-util");
var error_code_1 = require("../../error-code");
/**
 * Must be a valid date.
 *
 * `mixed.getTime()` must be a finite number.
 */
function instanceOfDate() {
    var unixTimestampMsDelegate = number_1.finiteNumber();
    return function (name, mixed) {
        if (!(mixed instanceof Date)) {
            throw error_util_1.makeMappingError({
                message: name + " must be instance of Date; received " + type_util_1.toTypeStr(mixed),
                inputName: name,
                actualValue: mixed,
                expected: "Date",
                expectedMeta: {
                    errorCode: error_code_1.ErrorCode.EXPECTED_TYPE,
                },
            });
        }
        var unixTimestampMs = mixed.getTime();
        unixTimestampMsDelegate(name + ".getTime()", unixTimestampMs);
        return mixed;
    };
}
exports.instanceOfDate = instanceOfDate;
