// type-mapping/dist/mapper/predicate/missing-value.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function isUndefinedMappable(f) {
    try {
        f("", undefined);
        return true;
    }
    catch (_err) {
        return false;
    }
}
exports.isUndefinedMappable = isUndefinedMappable;
function isNullMappable(f) {
    try {
        f("", null);
        return true;
    }
    catch (_err) {
        return false;
    }
}
exports.isNullMappable = isNullMappable;
function isMaybeMappable(f) {
    try {
        f("", null);
        f("", undefined);
        return true;
    }
    catch (_err) {
        return false;
    }
}
exports.isMaybeMappable = isMaybeMappable;
function canOutputUndefined(f) {
    try {
        /*
            Using idempotent property.

            (f(blah) === undefined) implies (f(undefined) === undefined)
        */
        return (f("", undefined) === undefined);
    }
    catch (_err) {
        return false;
    }
}
exports.canOutputUndefined = canOutputUndefined;
function canOutputNull(f) {
    try {
        /*
            Using idempotent property.

            (f(blah) === null) implies (f(null) === null)
        */
        return (f("", null) === null);
    }
    catch (_err) {
        return false;
    }
}
exports.canOutputNull = canOutputNull;
function canOutputMaybe(f) {
    try {
        /*
            Using idempotent property.

            (f(blah) === null) implies (f(null) === null)
            (f(blah) === undefined) implies (f(null) === undefined)
        */
        return ((f("", null) === null) &&
            (f("", undefined) === undefined));
    }
    catch (_err) {
        return false;
    }
}
exports.canOutputMaybe = canOutputMaybe;
