// type-mapping/dist/functional-lib/array-like/length.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var operator_1 = require("../operator");
var object_1 = require("../object");
var number_1 = require("../number");
var mapper_1 = require("../../mapper");
var string_1 = require("../string");
var operator_2 = require("../operator");
var string_util_1 = require("../../string-util");
var error_util_1 = require("../../error-util");
var error_code_1 = require("../../error-code");
function length(args) {
    var lengthDelegate = operator_1.pipe(number_1.unsignedInteger(), number_1.range({
        gtEq: args.min,
        ltEq: args.max,
    }));
    var expected = ((args.min == undefined) ?
        ((args.max == undefined) ?
            "value with \"length\" property" :
            "value of length less than, or equal to " + args.max.toString()) :
        ((args.max == undefined) ?
            "value of length greater than, or equal to " + args.min.toString() :
            (args.min == args.max) ?
                "value of length " + args.min.toString() :
                "value of length between " + args.min.toString() + " and " + args.max.toString()));
    return operator_1.pipe(operator_2.or(object_1.instanceOfObject(), string_1.string()), function (name, mixed) {
        var lengthResult = mapper_1.tryMapHandled(lengthDelegate, "" + name + string_util_1.toPropertyAccess("length"), mixed.length);
        if (lengthResult.success) {
            return mixed;
        }
        else {
            throw error_util_1.makeMappingError({
                message: name + " must be " + expected,
                inputName: name,
                actualValue: mixed,
                expected: expected,
                expectedMeta: {
                    errorCode: error_code_1.ErrorCode.EXPECTED_LENGTH,
                    min: args.min,
                    max: args.max,
                },
                propertyErrors: [
                    lengthResult.mappingError,
                ],
            });
        }
    });
}
exports.length = length;
