// type-mapping/dist/functional-lib/buffer/buffer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var type_util_1 = require("../../type-util");
var operator_1 = require("../operator");
var uint_8_array_1 = require("../uint-8-array");
var buffer_ctor_1 = require("../../type-util/buffer-ctor");
var error_util_1 = require("../../error-util");
var error_code_1 = require("../../error-code");
function instanceOfBuffer() {
    return function (name, mixed) {
        if (!buffer_ctor_1.isInstanceOfBuffer(mixed)) {
            throw error_util_1.makeMappingError({
                message: name + " must be instance of Buffer; received " + type_util_1.toTypeStr(mixed),
                inputName: name,
                actualValue: mixed,
                expected: "Buffer",
                expectedMeta: {
                    errorCode: error_code_1.ErrorCode.EXPECTED_TYPE,
                },
            });
        }
        return mixed;
    };
}
exports.instanceOfBuffer = instanceOfBuffer;
function bufferLength(args) {
    return operator_1.pipe(uint_8_array_1.byteLength(args), instanceOfBuffer());
}
exports.bufferLength = bufferLength;
function bufferExactLength(length) {
    return bufferLength({
        min: length,
        max: length,
    });
}
exports.bufferExactLength = bufferExactLength;
