// type-mapping/dist/fluent-lib/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
/// <reference path="../buffer.d.ts" />
var non_generic_1 = require("./non-generic");
exports.null = non_generic_1.null;
/**
    Export generated with `npm run generate-fluent-lib-export`

    The export generation script is needed because of this,
    https://github.com/microsoft/TypeScript/issues/31824
*/
var non_generic_2 = require("./non-generic");
exports.unsafeAny = non_generic_2.unsafeAny;
exports.instanceOfArray = non_generic_2.instanceOfArray;
exports.instanceOfReadOnlyArray = non_generic_2.instanceOfReadOnlyArray;
exports.implementsArrayLike = non_generic_2.implementsArrayLike;
exports.length = non_generic_2.length;
exports.bigInt = non_generic_2.bigInt;
exports.unsignedBigInt = non_generic_2.unsignedBigInt;
exports.stringToBigInt = non_generic_2.stringToBigInt;
exports.stringToUnsignedBigInt = non_generic_2.stringToUnsignedBigInt;
exports.integerToBigInt = non_generic_2.integerToBigInt;
exports.unsignedIntegerToUnsignedBigInt = non_generic_2.unsignedIntegerToUnsignedBigInt;
exports.toBigInt = non_generic_2.toBigInt;
exports.toUnsignedBigInt = non_generic_2.toUnsignedBigInt;
exports.bigIntGt = non_generic_2.bigIntGt;
exports.bigIntLt = non_generic_2.bigIntLt;
exports.bigIntGtEq = non_generic_2.bigIntGtEq;
exports.bigIntLtEq = non_generic_2.bigIntLtEq;
exports.bigIntRange = non_generic_2.bigIntRange;
exports.boolean = non_generic_2.boolean;
exports.finiteNumberToBoolean = non_generic_2.finiteNumberToBoolean;
exports.stringToBoolean = non_generic_2.stringToBoolean;
exports.finiteNumberToTrue = non_generic_2.finiteNumberToTrue;
exports.finiteNumberToFalse = non_generic_2.finiteNumberToFalse;
exports.stringToTrue = non_generic_2.stringToTrue;
exports.stringToFalse = non_generic_2.stringToFalse;
exports.toBoolean = non_generic_2.toBoolean;
exports.toTrue = non_generic_2.toTrue;
exports.toFalse = non_generic_2.toFalse;
exports.instanceOfBuffer = non_generic_2.instanceOfBuffer;
exports.bufferLength = non_generic_2.bufferLength;
exports.bufferExactLength = non_generic_2.bufferExactLength;
exports.unsafeStringToDate = non_generic_2.unsafeStringToDate;
exports.dateToString = non_generic_2.dateToString;
exports.unixTimestampSecondsToDate = non_generic_2.unixTimestampSecondsToDate;
exports.unixTimestampMillisecondsToDate = non_generic_2.unixTimestampMillisecondsToDate;
exports.dateToUnixTimestampSeconds = non_generic_2.dateToUnixTimestampSeconds;
exports.dateToUnixTimestampMilliseconds = non_generic_2.dateToUnixTimestampMilliseconds;
exports.instanceOfDate = non_generic_2.instanceOfDate;
exports.stringToJsonObject = non_generic_2.stringToJsonObject;
exports.jsonObject = non_generic_2.jsonObject;
exports.never = non_generic_2.never;
exports.nonMaybe = non_generic_2.nonMaybe;
exports.undefinedToNull = non_generic_2.undefinedToNull;
exports.emptyStringToNull = non_generic_2.emptyStringToNull;
exports.whitespaceStringToNull = non_generic_2.whitespaceStringToNull;
exports.stringToFiniteNumber = non_generic_2.stringToFiniteNumber;
exports.stringToInteger = non_generic_2.stringToInteger;
exports.stringToUnsignedInteger = non_generic_2.stringToUnsignedInteger;
exports.toFiniteNumber = non_generic_2.toFiniteNumber;
exports.toInteger = non_generic_2.toInteger;
exports.toUnsignedInteger = non_generic_2.toUnsignedInteger;
exports.gt = non_generic_2.gt;
exports.lt = non_generic_2.lt;
exports.gtEq = non_generic_2.gtEq;
exports.ltEq = non_generic_2.ltEq;
exports.range = non_generic_2.range;
exports.unsafeNumber = non_generic_2.unsafeNumber;
exports.finiteNumber = non_generic_2.finiteNumber;
exports.integer = non_generic_2.integer;
exports.unsignedInteger = non_generic_2.unsignedInteger;
exports.toEmptyObject = non_generic_2.toEmptyObject;
exports.instanceOfObject = non_generic_2.instanceOfObject;
exports.pojo = non_generic_2.pojo;
exports.runTimeRequired = non_generic_2.runTimeRequired;
exports.notRunTimeRequired = non_generic_2.notRunTimeRequired;
exports.finiteNumberToFiniteNumberString = non_generic_2.finiteNumberToFiniteNumberString;
exports.integerToIntegerString = non_generic_2.integerToIntegerString;
exports.unsignedIntegerToUnsignedIntegerString = non_generic_2.unsignedIntegerToUnsignedIntegerString;
exports.jsonObjectToJsonObjectString = non_generic_2.jsonObjectToJsonObjectString;
exports.ipAddressString = non_generic_2.ipAddressString;
exports.ipV4MappedIpV6String = non_generic_2.ipV4MappedIpV6String;
exports.ipV4OctetString = non_generic_2.ipV4OctetString;
exports.ipV4String = non_generic_2.ipV4String;
exports.ipV6SegmentString = non_generic_2.ipV6SegmentString;
exports.ipV6StringWithMaxSegmentCount = non_generic_2.ipV6StringWithMaxSegmentCount;
exports.ipV6String = non_generic_2.ipV6String;
exports.floatingPointFormatString = non_generic_2.floatingPointFormatString;
exports.integerFormatString = non_generic_2.integerFormatString;
exports.unsignedIntegerFormatString = non_generic_2.unsignedIntegerFormatString;
exports.finiteNumberString = non_generic_2.finiteNumberString;
exports.integerString = non_generic_2.integerString;
exports.unsignedIntegerString = non_generic_2.unsignedIntegerString;
exports.string = non_generic_2.string;
exports.jsonObjectString = non_generic_2.jsonObjectString;
exports.stringLength = non_generic_2.stringLength;
exports.stringExactLength = non_generic_2.stringExactLength;
exports.match = non_generic_2.match;
exports.notMatch = non_generic_2.notMatch;
exports.email = non_generic_2.email;
exports.emailAddress = non_generic_2.emailAddress;
exports.hexadecimalString = non_generic_2.hexadecimalString;
exports.toUpperCase = non_generic_2.toUpperCase;
exports.toLowerCase = non_generic_2.toLowerCase;
exports.padLeft = non_generic_2.padLeft;
exports.padRight = non_generic_2.padRight;
exports.subStringBlacklist = non_generic_2.subStringBlacklist;
exports.toTrimmed = non_generic_2.toTrimmed;
exports.byteLength = non_generic_2.byteLength;
exports.instanceOfUint8Array = non_generic_2.instanceOfUint8Array;
exports.uint8ArrayLength = non_generic_2.uint8ArrayLength;
exports.uint8ArrayExactLength = non_generic_2.uint8ArrayExactLength;
exports.nullToUndefined = non_generic_2.nullToUndefined;
exports.emptyStringToUndefined = non_generic_2.emptyStringToUndefined;
exports.whitespaceStringToUndefined = non_generic_2.whitespaceStringToUndefined;
exports.undefined = non_generic_2.undefined;
exports.unknown = non_generic_2.unknown;
__export(require("./generic"));
__export(require("./field-map-ctor"));
var ArrayBufferUtil = require("../array-buffer-util");
exports.ArrayBufferUtil = ArrayBufferUtil;
var EnumUtil = require("../enum-util");
exports.EnumUtil = EnumUtil;
var BigIntUtil = require("../bigint-util");
exports.BigIntUtil = BigIntUtil;
var FixedPointUtil = require("../fixed-point-util");
exports.FixedPointUtil = FixedPointUtil;
var FloatingPointUtil = require("../floating-point-util");
exports.FloatingPointUtil = FloatingPointUtil;
__export(require("../decorator"));
var ErrorUtil = require("../error-util");
exports.ErrorUtil = ErrorUtil;
var jsonApi = require("../json-api-lib");
exports.jsonApi = jsonApi;
__export(require("../mapper"));
var mysql = require("../mysql-lib");
exports.mysql = mysql;
var TypeUtil = require("../type-util");
exports.TypeUtil = TypeUtil;
__export(require("../error-code"));
__export(require("../fluent-mapper"));
