// type-mapping/dist/functional-lib/object/rename.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var mapper_1 = require("../../mapper");
var operator_1 = require("../operator");
var instance_of_object_1 = require("./instance-of-object");
var string_util_1 = require("../../string-util");
var error_util_1 = require("../../error-util");
function rename(srcKey, dstKey, f) {
    var runTimeRequired = mapper_1.getRunTimeRequiredFlagOrFalse(f);
    var result = operator_1.pipe(instance_of_object_1.instanceOfObject(), function (name, mixed) {
        var _a;
        var unsafeKey = "";
        var unsafeName = "";
        var unsafeValue = undefined;
        if (Object.prototype.hasOwnProperty.call(mixed, dstKey)) {
            unsafeKey = dstKey;
            unsafeName = "" + name + string_util_1.toPropertyAccess(dstKey);
            unsafeValue = mixed[dstKey];
        }
        else if (Object.prototype.hasOwnProperty.call(mixed, srcKey)) {
            unsafeKey = srcKey;
            unsafeName = "" + name + string_util_1.toPropertyAccess(srcKey);
            unsafeValue = mixed[srcKey];
        }
        else if (runTimeRequired) {
            var expected = "object with explicitly set property " + JSON.stringify(dstKey);
            throw error_util_1.makeMappingError({
                message: name + " must be " + expected,
                inputName: name,
                actualValue: mixed,
                expected: expected,
                expectedMeta: {
                    srcKey: srcKey,
                    dstKey: dstKey,
                },
                propertyErrors: [
                    error_util_1.makeMappingError({
                        message: "" + name + string_util_1.toPropertyAccess(dstKey) + " must be explicitly set",
                        inputName: "" + name + string_util_1.toPropertyAccess(dstKey),
                        actualValue: undefined,
                        expected: "explicitly set",
                    }),
                ],
            });
        }
        else {
            unsafeKey = dstKey;
            unsafeName = "" + name + string_util_1.toPropertyAccess(dstKey);
            unsafeValue = undefined;
        }
        var dstResult = mapper_1.tryMapHandled(f, unsafeName, unsafeValue);
        if (dstResult.success) {
            var obj = (_a = {},
                _a[dstKey] = dstResult.value,
                _a);
            return obj;
        }
        if (dstResult.mappingError.expected == undefined) {
            throw error_util_1.makeMappingError({
                message: name + " must be valid object. " + dstResult.mappingError.message,
                inputName: name,
                actualValue: mixed,
                expected: "valid object",
                expectedMeta: {
                    srcKey: srcKey,
                    dstKey: dstKey,
                },
                propertyErrors: [
                    dstResult.mappingError,
                ],
            });
        }
        else {
            var expected = "object with " + dstResult.mappingError.expected + " property " + JSON.stringify(unsafeKey);
            throw error_util_1.makeMappingError({
                message: name + " must be " + expected,
                inputName: name,
                actualValue: mixed,
                expected: expected,
                expectedMeta: {
                    srcKey: srcKey,
                    dstKey: dstKey,
                },
                propertyErrors: [
                    dstResult.mappingError,
                ],
            });
        }
    });
    return mapper_1.copyRunTimeModifier(f, result);
}
exports.rename = rename;
/*
const a = unsignedInteger();
const b = optional(unsignedInteger());
const c = string();
const d = optional(string());

declare const _0 : typeof a;
declare const _1 : typeof b;
declare const _2 : typeof c;
declare const _3 : typeof d;

declare const _01 : (typeof a) | (typeof b);
declare const _02 : (typeof a) | (typeof c);
declare const _03 : (typeof a) | (typeof d);

declare const _12 : (typeof b) | (typeof c);
declare const _13 : (typeof b) | (typeof d);

declare const _23 : (typeof c) | (typeof d);

const __0 = rename("x", "y", _0);
const __1 = rename("x", "y", _1);
const __2 = rename("x", "y", _2);
const __3 = rename("x", "y", _3);

const __01 = rename("x", "y", _01);
const __02 = rename("x", "y", _02);
const __03 = rename("x", "y", _03);

const __12 = rename("x", "y", _12);
const __13 = rename("x", "y", _13);

const __23 = rename("x", "y", _23);
*/
