// type-mapping/dist/mapper/operation/try-map.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var error_util_1 = require("../../error-util");
function tryMapImpl(f, name, mixed) {
    try {
        return {
            success: true,
            value: f(name, mixed),
        };
    }
    catch (err) {
        if (error_util_1.isMappingError(err)) {
            return {
                success: false,
                err: err,
                mappingError: err,
            };
        }
        else {
            return {
                success: false,
                err: err,
                mappingError: error_util_1.makeMappingError({
                    message: err.message,
                    inputName: name,
                    actualValue: mixed,
                    expected: undefined,
                    stack: err.stack,
                }),
            };
        }
    }
}
/**
    Alias of `tryMapExpected<>()`
*/
function tryMap(f, name, mixed) {
    return tryMapImpl(f, name, mixed);
}
exports.tryMap = tryMap;
;
function tryMapExpected(f, name, mixed) {
    return tryMapImpl(f, name, mixed);
}
exports.tryMapExpected = tryMapExpected;
;
function tryMapMappable(f, name, mixed) {
    return tryMapImpl(f, name, mixed);
}
exports.tryMapMappable = tryMapMappable;
;
function tryMapHandled(f, name, mixed) {
    return tryMapImpl(f, name, mixed);
}
exports.tryMapHandled = tryMapHandled;
;
