// type-mapping/dist/mysql-lib/int/int.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fLib = require("../../fluent-lib");
var functional_lib_1 = require("../../functional-lib");
var type_util_1 = require("../../type-util");
/**
    Uses `integerFormatString()` and `BigInt()` internally.
*/
function unsafeInt() {
    return fLib.toBigInt();
}
exports.unsafeInt = unsafeInt;
function intDelegate(min, max) {
    var bigIntFactory = type_util_1.getBigIntFactoryFunctionOrError();
    return unsafeInt().pipe(functional_lib_1.bigIntRange({
        gtEq: bigIntFactory(min),
        ltEq: bigIntFactory(max),
    }));
}
/**
    [-128, 127]
*/
function tinyIntSigned() {
    return intDelegate(-128, 127);
}
exports.tinyIntSigned = tinyIntSigned;
/**
    [-32,768, 32,767]
*/
function smallIntSigned() {
    return intDelegate(-32768, 32767);
}
exports.smallIntSigned = smallIntSigned;
/**
    [-8,388,608, 8,388,607]
*/
function mediumIntSigned() {
    return intDelegate(-8388608, 8388607);
}
exports.mediumIntSigned = mediumIntSigned;
/**
    [-2,147,483,648, 2,147,483,647]
*/
function intSigned() {
    return intDelegate(-2147483648, 2147483647);
}
exports.intSigned = intSigned;
/**
    [0, 255]
*/
function tinyIntUnsigned() {
    return intDelegate(0, 255);
}
exports.tinyIntUnsigned = tinyIntUnsigned;
/**
    [0, 65535]
*/
function smallIntUnsigned() {
    return intDelegate(0, 65535);
}
exports.smallIntUnsigned = smallIntUnsigned;
/**
    [0, 16777215]
*/
function mediumIntUnsigned() {
    return intDelegate(0, 16777215);
}
exports.mediumIntUnsigned = mediumIntUnsigned;
/**
    [0, 4294967295]
*/
function intUnsigned() {
    return intDelegate(0, 4294967295);
}
exports.intUnsigned = intUnsigned;
