// type-mapping/dist/mysql-lib/binary/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fluent_lib_1 = require("../../fluent-lib");
function bufferDelegate(dataTypeStr, maxLength) {
    var result = function (a, b) {
        if (a == undefined) {
            return fluent_lib_1.bufferLength({
                max: maxLength,
            });
        }
        else if (b == undefined) {
            a = fluent_lib_1.pipe(fluent_lib_1.integer(), fluent_lib_1.range({
                gtEq: 1,
                ltEq: maxLength,
            }))(dataTypeStr + ".desiredLengthMax", a);
            return fluent_lib_1.bufferLength({
                max: a,
            });
        }
        else {
            a = fluent_lib_1.pipe(fluent_lib_1.integer(), fluent_lib_1.range({
                gtEq: 0,
                ltEq: maxLength,
            }))(dataTypeStr + ".desiredLengthMin", a);
            b = fluent_lib_1.pipe(fluent_lib_1.integer(), fluent_lib_1.range({
                gtEq: 1,
                ltEq: maxLength,
            }))(dataTypeStr + ".desiredLengthMax", b);
            if (a > b) {
                throw new Error(dataTypeStr + " minLength must be <= maxLength");
            }
            return fluent_lib_1.bufferLength({
                min: a,
                max: b,
            });
        }
    };
    result.maxLength = maxLength;
    return result;
}
exports.binary = bufferDelegate("BINARY", 255);
exports.varBinary = bufferDelegate("VARBINARY", 65535);
exports.tinyBlob = bufferDelegate("TINYBLOB", 255);
exports.blob = bufferDelegate("BLOB", 65535);
exports.mediumBlob = bufferDelegate("MEDIUMBLOB", 16777215);
exports.longBlob = bufferDelegate("LONGBLOB", 4294967295);
