// type-mapping/dist/mysql-lib/uint-8-array-binary/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fluent_lib_1 = require("../../fluent-lib");
function uint8ArrayDelegate(dataTypeStr, maxLength) {
    var result = function (a, b) {
        if (a == undefined) {
            return fluent_lib_1.uint8ArrayLength({
                max: maxLength,
            });
        }
        else if (b == undefined) {
            a = fluent_lib_1.pipe(fluent_lib_1.integer(), fluent_lib_1.range({
                gtEq: 1,
                ltEq: maxLength,
            }))(dataTypeStr + ".desiredLengthMax", a);
            return fluent_lib_1.uint8ArrayLength({
                max: a,
            });
        }
        else {
            a = fluent_lib_1.pipe(fluent_lib_1.integer(), fluent_lib_1.range({
                gtEq: 0,
                ltEq: maxLength,
            }))(dataTypeStr + ".desiredLengthMin", a);
            b = fluent_lib_1.pipe(fluent_lib_1.integer(), fluent_lib_1.range({
                gtEq: 1,
                ltEq: maxLength,
            }))(dataTypeStr + ".desiredLengthMax", b);
            if (a > b) {
                throw new Error(dataTypeStr + " minLength must be <= maxLength");
            }
            return fluent_lib_1.uint8ArrayLength({
                min: a,
                max: b,
            });
        }
    };
    result.maxLength = maxLength;
    return result;
}
exports.uint8ArrayBinary = uint8ArrayDelegate("BINARY", 255);
exports.uint8ArrayVarBinary = uint8ArrayDelegate("VARBINARY", 65535);
exports.uint8ArrayTinyBlob = uint8ArrayDelegate("TINYBLOB", 255);
exports.uint8ArrayBlob = uint8ArrayDelegate("BLOB", 65535);
exports.uint8ArrayMediumBlob = uint8ArrayDelegate("MEDIUMBLOB", 16777215);
exports.uint8ArrayLongBlob = uint8ArrayDelegate("LONGBLOB", 4294967295);
