// type-mapping/dist/functional-lib/bigint/bigint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var type_util_1 = require("../../type-util");
var operator_1 = require("../operator");
var comparison_1 = require("./comparison");
var error_util_1 = require("../../error-util");
var error_code_1 = require("../../error-code");
function bigInt() {
    return function (name, mixed) {
        if (type_util_1.isBigInt(mixed)) {
            return mixed;
        }
        throw error_util_1.makeMappingError({
            message: name + " must be bigint; received " + type_util_1.toTypeStr(mixed),
            inputName: name,
            actualValue: mixed,
            expected: "bigint",
            expectedMeta: {
                errorCode: error_code_1.ErrorCode.EXPECTED_TYPE,
            },
        });
    };
}
exports.bigInt = bigInt;
/**
    bigint >= 0
*/
function unsignedBigInt() {
    var bigIntFactory = type_util_1.getBigIntFactoryFunctionOrError();
    return operator_1.pipe(bigInt(), comparison_1.bigIntGtEq(bigIntFactory(0)));
}
exports.unsignedBigInt = unsignedBigInt;
