// type-mapping/dist/functional-lib/bigint/cast.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var string_1 = require("../string");
var operator_1 = require("../operator");
var bigint_1 = require("./bigint");
var type_util_1 = require("../../type-util");
var number_1 = require("../number");
/**
    Uses `integerFormatString()` and `BigInt()` internally.
*/
function stringToBigInt() {
    var bigIntFactory = type_util_1.getBigIntFactoryFunctionOrError();
    return operator_1.cast(string_1.integerFormatString(), bigIntFactory, bigint_1.bigInt());
}
exports.stringToBigInt = stringToBigInt;
/**
    Uses `unsignedIntegerFormatString()` and `BigInt()` internally.
*/
function stringToUnsignedBigInt() {
    var bigIntFactory = type_util_1.getBigIntFactoryFunctionOrError();
    return operator_1.cast(string_1.unsignedIntegerFormatString(), bigIntFactory, bigint_1.unsignedBigInt());
}
exports.stringToUnsignedBigInt = stringToUnsignedBigInt;
function integerToBigInt() {
    var bigIntFactory = type_util_1.getBigIntFactoryFunctionOrError();
    return operator_1.cast(number_1.integer(), bigIntFactory, bigint_1.bigInt());
}
exports.integerToBigInt = integerToBigInt;
function unsignedIntegerToUnsignedBigInt() {
    var bigIntFactory = type_util_1.getBigIntFactoryFunctionOrError();
    return operator_1.cast(number_1.unsignedInteger(), bigIntFactory, bigint_1.unsignedBigInt());
}
exports.unsignedIntegerToUnsignedBigInt = unsignedIntegerToUnsignedBigInt;
/**
    Uses `integerFormatString()` and `BigInt()` internally.
*/
function toBigInt() {
    var bigIntFactory = type_util_1.getBigIntFactoryFunctionOrError();
    return operator_1.cast(operator_1.or(string_1.integerFormatString(), number_1.integer()), bigIntFactory, bigint_1.bigInt());
}
exports.toBigInt = toBigInt;
/**
    Uses `unsignedIntegerFormatString()` and `BigInt()` internally.
*/
function toUnsignedBigInt() {
    var bigIntFactory = type_util_1.getBigIntFactoryFunctionOrError();
    return operator_1.cast(operator_1.or(string_1.unsignedIntegerFormatString(), number_1.unsignedInteger()), bigIntFactory, bigint_1.unsignedBigInt());
}
exports.toUnsignedBigInt = toUnsignedBigInt;
