// type-mapping/dist/functional-lib/uint-8-array/uint-8-array.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var type_util_1 = require("../../type-util");
var operator_1 = require("../operator");
var byte_length_1 = require("./byte-length");
var error_util_1 = require("../../error-util");
var error_code_1 = require("../../error-code");
function instanceOfUint8Array() {
    return function (name, mixed) {
        if (!(mixed instanceof Uint8Array)) {
            throw error_util_1.makeMappingError({
                message: name + " must be instance of Uint8Array; received " + type_util_1.toTypeStr(mixed),
                inputName: name,
                actualValue: mixed,
                expected: "Uint8Array",
                expectedMeta: {
                    errorCode: error_code_1.ErrorCode.EXPECTED_TYPE,
                },
            });
        }
        return mixed;
    };
}
exports.instanceOfUint8Array = instanceOfUint8Array;
function uint8ArrayLength(args) {
    return operator_1.pipe(byte_length_1.byteLength(args), instanceOfUint8Array());
}
exports.uint8ArrayLength = uint8ArrayLength;
function uint8ArrayExactLength(length) {
    return uint8ArrayLength({
        min: length,
        max: length,
    });
}
exports.uint8ArrayExactLength = uint8ArrayExactLength;
