// type-mapping/dist/mysql-lib/char/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fluent_lib_1 = require("../../fluent-lib");
function stringDelegate(dataTypeStr, maxLength) {
    var result = function (a, b) {
        if (a == undefined) {
            return fluent_lib_1.stringLength({
                max: maxLength,
            });
        }
        else if (b == undefined) {
            a = fluent_lib_1.pipe(fluent_lib_1.integer(), fluent_lib_1.range({
                gtEq: 1,
                ltEq: maxLength,
            }))(dataTypeStr + ".desiredLengthMax", a);
            return fluent_lib_1.stringLength({
                max: a,
            });
        }
        else {
            a = fluent_lib_1.pipe(fluent_lib_1.integer(), fluent_lib_1.range({
                gtEq: 0,
                ltEq: maxLength,
            }))(dataTypeStr + ".desiredLengthMin", a);
            b = fluent_lib_1.pipe(fluent_lib_1.integer(), fluent_lib_1.range({
                gtEq: 1,
                ltEq: maxLength,
            }))(dataTypeStr + ".desiredLengthMax", b);
            if (a > b) {
                throw new Error(dataTypeStr + " minLength must be <= maxLength");
            }
            return fluent_lib_1.stringLength({
                min: a,
                max: b,
            });
        }
    };
    result.maxLength = maxLength;
    return result;
}
exports.char = stringDelegate("CHAR", 255);
exports.varChar = stringDelegate("VARCHAR", 65535);
exports.tinyText = stringDelegate("TINYTEXT", 255);
exports.text = stringDelegate("TEXT", 65535);
exports.mediumText = stringDelegate("MEDIUMTEXT", 16777215);
exports.longText = stringDelegate("LONGTEXT", 4294967295);
