// type-mapping/dist/json-api-lib/client-resource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fLib = require("../fluent-lib");
var attribute_collection_1 = require("./attribute-collection");
var relationship_1 = require("./relationship");
var link_collection_1 = require("./link-collection");
var meta_1 = require("./meta");
/**
 * https://jsonapi.org/format/1.0/#document-resource-objects
 *
 * “Resource objects” appear in a JSON:API document to represent resources.
 *
 * A resource object **MUST** contain at least the following top-level members:
 *
 * + `id`
 * + `type`
 *
 * Exception: The `id` member is not required when the resource object originates
 * at the client and represents a new resource to be created on the server.
 *
 * In addition, a resource object **MAY** contain any of these top-level members:
 *
 * + `attributes`: an attributes object representing some of the resource’s data.
 * + `relationships`: a relationships object describing relationships between the resource and other JSON:API resources.
 * + `links`: a links object containing links related to the resource.
 * + `meta`: a meta object containing non-standard meta-information about a resource that can not be
 *   represented as an attribute or relationship.
 *
 *  @see {@link attributeCollection}
 *  @see {@link relationship}
 *  @see {@link linkCollection}
 *  @see {@link meta}
 *  @see {@link ClientResource}
 *  @see {@link ExpectedInputClientResource}
 */
exports.clientResource = function () { return fLib.object({
    id: fLib.string().orNull().optional(),
    type: fLib.string(),
    attributes: attribute_collection_1.attributeCollection().orNull().optional(),
    relationships: relationship_1.relationship().orNull().optional(),
    links: link_collection_1.linkCollection().orNull().optional(),
    meta: meta_1.meta().orNull().optional(),
}); };
