// type-mapping/dist/functional-lib/operator/missing-value.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var mapper_1 = require("../../mapper");
var or_1 = require("./or");
var literal_1 = require("../literal");
var exclude_literal_1 = require("./exclude-literal");
function orUndefined(f) {
    return or_1.or(f, literal_1.literal(undefined));
}
exports.orUndefined = orUndefined;
function orNull(f) {
    return or_1.or(f, literal_1.literal(null));
}
exports.orNull = orNull;
function orMaybe(f) {
    return or_1.or(f, literal_1.literal(undefined), literal_1.literal(null));
}
exports.orMaybe = orMaybe;
function notUndefined(f) {
    return exclude_literal_1.excludeLiteral(f, undefined);
}
exports.notUndefined = notUndefined;
function notNull(f) {
    return exclude_literal_1.excludeLiteral(f, null);
}
exports.notNull = notNull;
function notMaybe(assert) {
    return exclude_literal_1.excludeLiteral(assert, undefined, null);
}
exports.notMaybe = notMaybe;
function optional(f) {
    var g = orUndefined(f);
    g.__optional = true;
    g.__runTimeRequired = false;
    return g;
}
exports.optional = optional;
;
function notOptional(f) {
    var g = notUndefined(f);
    g.__optional = false;
    return g;
}
exports.notOptional = notOptional;
;
function runTimeRequired(f) {
    var g = function (name, mixed) {
        return f(name, mixed);
    };
    g.__runTimeRequired = true;
    g.__optional = false;
    return mapper_1.copyName(f, g);
}
exports.runTimeRequired = runTimeRequired;
;
function notRunTimeRequired(f) {
    var g = function (name, mixed) {
        return f(name, mixed);
    };
    g.__runTimeRequired = false;
    return mapper_1.copyName(f, mapper_1.copyOptional(f, g));
}
exports.notRunTimeRequired = notRunTimeRequired;
;
