// type-mapping/dist/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
/// <reference path="buffer.d.ts" />
var ArrayBufferUtil = require("./array-buffer-util");
exports.ArrayBufferUtil = ArrayBufferUtil;
var EnumUtil = require("./enum-util");
exports.EnumUtil = EnumUtil;
var BigIntUtil = require("./bigint-util");
exports.BigIntUtil = BigIntUtil;
var FixedPointUtil = require("./fixed-point-util");
exports.FixedPointUtil = FixedPointUtil;
var FloatingPointUtil = require("./floating-point-util");
exports.FloatingPointUtil = FloatingPointUtil;
__export(require("./decorator"));
var ErrorUtil = require("./error-util");
exports.ErrorUtil = ErrorUtil;
__export(require("./field-map"));
__export(require("./functional-lib"));
var jsonApi = require("./json-api-lib");
exports.jsonApi = jsonApi;
__export(require("./mapper"));
var mysql = require("./mysql-lib");
exports.mysql = mysql;
var TypeUtil = require("./type-util");
exports.TypeUtil = TypeUtil;
__export(require("./error-code"));
__export(require("./fluent-mapper"));
/**
 * Convenience function.
 * Will create a native `bigint` (if natively supported),
 * or create a polyfilled `bigint` (if polyfilled).
 */
function BigInt(x) {
    return TypeUtil.getBigIntFactoryFunctionOrError()(x);
}
exports.BigInt = BigInt;
